### DISCUSSION TITLE
Force cache refresh before expiration?

### DISCUSSION THREAD

**Discussion by kumy** [dimo414/bash-cache#28] [Q&A] [answered]
@dimo414 is there a way to force a decorated function to run again it's original function, return the fresh result and refresh the cache before it has expired? Or is there a way to mark the cached version as expired so the original function will run again?

**Comment by dimo414** [comment, marked as answer]
Can you share more details about what you're trying to do?

You can use an environment variable to invalidate the cache (`PWD` is particularly useful so commands run in different directories are cached separately) or `bc::warm::FUNCTION_NAME` to asynchronously warm the cache. You can also customize the `BC_CACHE_DIR` to invalidate old cached data, though that's a bit of a big hammer. There isn't a way to force a single command to be re-cached at present ([`bkt`](http://github.com/dimo414/bkt) provides a `--force` flag that does this) but I can imagine adding a `bc::force::FUNCTION_NAME` function if that would be useful.

**Reply by kumy** [reply]
I don't think changing the ENVs will help in our context :/, as of pruning the whole `BC_CACHE_DIR` which seems a bit like a bazooka!

`bc::force::FUNCTION_NAME` would be super nice, as long as it returns "fresh" data and set the cache appropriately for next "normal" call.

One of the use case:
we have 4 folders/environments, in each of them a very expensive function will have it's return cached (with `PWD` as key) for 24h

But sometime we know one of the environment has changes, and we would like the cache for it to be invalidated, so the function will return fresh results. All other env must keep they cache.

**Reply by kumy** [reply]
I thought, a call to an expire function `bc::expire::FUNCTION_NAME` (from script A) to mark it as expired, then next call (from script B) on `FUNCTION_NAME` should not use the cache, and return fresh output would be a good pattern for us.

**Reply by dimo414** [reply]
If you haven't already, do take a look at [`bkt`](https://github.com/dimo414/bkt), it's similar to bash-cache and supports `--force`-ing a cache miss. I've been considering more powerful invalidation mechanisms in https://github.com/dimo414/bkt/issues/5 and https://github.com/dimo414/bkt/issues/10 as well, and would appreciate any opinions on either of those issues.

> sometime we know one of the environment has changes

Can you elaborate on this, e.g. what signals do you use to know there have been changes? I'm definitely open to adding `bc::force::FUNCTION_NAME` but I'd like to understand if there are other things we could do to make the cache key smarter.

**Reply by kumy** [reply]
thanks for your answer. In our project we try to restrict at the bare minimum the external dependencies the user has to install on it computer to use it. Adding a new binary (bkt) is not an option (yet).


>   sometime we know one of the environment has changes
> 
> Can you elaborate on this, e.g. what signals do you use to know there have been changes?

- We have a command (think about it as a ~dashboard) that display some values from k8s, like the image used in a deployment. That run in a terminal in a never ending loop.
- That "dashboard" call a `k8sget` function (`function k8sget() { kubectl get ... -o json; } && bc::cache k8sget ...`) and pretty display it. That function output is put in cache for a long period thanks to `bc-cache`
- Then from time to time, we do a new product release, the image change in k8s. We (human) know that, and we were used to launch another command, which does `rm .cache` (from our old template system - before switching to bc-cache). The "dashboard: still running on another terminal, will on it's next loop call `k8sget` and we expect to see a fresh value.

- having a `bc::**expire**::FUNCTION_NAME` function: we could replace the `rm` in the clean cache script by this easily. Command will return **immediately**.
- having a`bc::**force**::FUNCTION_NAME` function could be the same but the command will take longer to terminate and output nothing as we don't care about the result at that point.

Does it make sense?

**Reply by dimo414** [reply]
Thanks yes, that's very helpful. Is there a reason you don't want to use `bc::warm::k8sget` to trigger the refresh? The old value would continue to display on the dashboard until the function completes, but it sounds like that wouldn't be an issue in this case?

I don't like `bc::expire::FUNCTION_NAME` as much because it's less obvious how that works for functions that can take different arguments (whereas `warm`/`force` executes exactly the command given), but IIUC warming would do what you need, since you're trying to refresh the cache not just clear it.

**Reply by kumy** [reply]
From the [readme](https://github.com/dimo414/bash-cache#warming-the-cache) I didn't understood the warm will force the cache to refresh. I thought that it would use the cache if it's still valid. I will take a look a it.

**Reply by kumy** [reply]
ok, `bc::warm::k8sget` may work. But as the call will trigger the refresh directly, we may want to do some other stuff after the command finished, like before giving the hand to the user output some text using the values we just reloaded. Any way to have this call blocking? I tried to  follow it by `wait` but that didn't helped. `bc::**force**::FUNCTION_NAME` might help for this case.

**Reply by dimo414** [reply]
Yeah `warm` is intentionally async, the intent is to be able to initiate a refresh before the data is needed. If you need to block until completion then we'll want to add a `force` operation. It might be possible to get `wait` to work, but I haven't tried it and it's not very user-friendly in any case.

**Reply by dimo414** [reply]
I've gone ahead and added `bc::force::FUNCTION_NAME` and re-written that section of the README. Feel free to send a PR with further edits to the README if you think it could still be clearer.

**Reply by kumy** [reply]
Will try that! Many thanks!! :partying_face:

**Reply by kumy** [reply]
The README looks fine, thanks.

**Reply by kumy** [reply]
Just to confirm that new `bc::force::FUNCTION_NAME` works like a charm! Thanks
